OkHttp optimizations

pupitetris opened the pull request:
This pull request allows the reuse of the okHttpClient object, as recommended by the library's documentation.

* okHttpClient reuse allows for HTTP/2 connecton reuse and reduces
overhead.

* sets eager timeout specs to circumvent Roku issues regarding a lack
of priority for the ECP web server and not handling requests promptly
enough when the appliance is busy, such as with fast browsing of
collections or search results or while gaming. A low timeout cancels
requests that are not being timely handled, eluding "catch up"
interactions that only generate out-of-sequence inputs.

* sets up a connection pool with a 1 sec keepalive so HTTP/2
connections can be reused and to try to keep the ECP web server from
sleeping.

* just for a tiny bit of performance, a quick request object cache is
put in place based on URL, for requests where lag is relevant (those
being key press/up/down requests).

pupitetris committed:
ecp/core/HttpManager: new singleton for better okHttpClient use

allows the reuse of the okHttpClient object, as recommended by the
library's documentation.

* okHttpClient reuse allows for HTTP/2 connecton reuse and reduces
overhead.

* sets eager timeout specs to circumvent Roku issues regarding a lack
of priority for the ECP web server and not handling requests promptly
enough when the appliance is busy, such as with fast browsing of
collections or search results or while gaming. A low timeout cancels
requests that are not being timely handled, eluding "catch up"
interactions that only generate out-of-sequence inputs.

* sets up a connection pool with a 1 sec keepalive so HTTP/2
connections can be reused and to try to keep the ECP web server from
sleeping.

* just for a tiny bit of performance, a quick request object cache is
put in place based on URL, for requests where lag is relevant (those
being key press/up/down requests).

pupitetris committed:
ecp/core/ECPRequest: use HttpManager when sending requests

also, new private member useRequestCache specifies if the Request
should tell HttpManager to cache the okHttp.Request object.

pupitetris committed:
ecp/request/Key*.java: use HttpManager's RequestCache

wseemannroku commented on src/main/java/com/wseemann/ecp/core/HttpManager.kt:
```diff
@@ -0,0 +1,57 @@
+import okhttp3.Request
+import okhttp3.RequestBody.Companion.toRequestBody
+
+class HttpManager private constructor() {
```
Can you mark this as `internal class` please? It shouldn't be exposed as an end user API.

wseemann replied:
Can you mark this as internal class please? It shouldn't be exposed as an end user API.

wseemannroku commented on src/main/java/com/wseemann/ecp/core/HttpManager.kt:
```diff
@@ -0,0 +1,57 @@
+
+    private val okHttpClient = OkHttpClient.Builder()
+	.connectionPool(ConnectionPool(20, 1, TimeUnit.SECONDS))
+	.connectTimeout(40, TimeUnit.MILLISECONDS)
```
These are extremely low connection and read times. I'd prefer at least two seconds. While these times may work on your internet connection other users may have less than ideal network conditions are setting the values this low would make this SDK unusable.

wseemann commented on src/main/java/com/wseemann/ecp/core/HttpManager.kt:
```diff
@@ -0,0 +1,57 @@
+
+class HttpManager private constructor() {
+
+    private val okHttpClient = OkHttpClient.Builder()
```
These are extremely low connection and read times. I'd prefer at least two seconds. While these times may work on your internet connection other users may have less than ideal network conditions are setting the values this low would make this SDK unusable.

pupitetris committed:
ECPRequest: keep and reuse the generated okHTTP Request

This allows for the ECPRequest to be reused at the application level
without having to rebuild the okHttp Request every time.

HttpManager: improved public methods granularity to expose the
okHttp Request.

pupitetris committed:
HttpManager: class marked as internal

pupitetris committed:
HttpManager: lifted connection timeouts to 2 seconds

pupitetris commented:
OK, both modifications are in.

Since ECP key requests don't change at the HTTP level, I also added that the okHTTP requests are kept in memory once they are built. This way, if the ECPRequest is not destroyed by the application code, the okHTTP request needs not be rebuilt when the request is sent again.

wseemann committed:
Fix formatting

wseemann commented on src/main/java/com/wseemann/ecp/core/ECPRequest.kt:
```diff
@@ -33,31 +38,16 @@ internal abstract class ECPRequest<T>(private val url: String) {
-                    return null
+                val httpManager = HttpManager.getInstance()
+
+                if (http_request == null) {
```
Can you explain the purpose of `http_request` here? Wouldn't this cause caching of every request regardless of the `useRequestCache` flag because once the value is set to `true` then:

 http_request = httpManager.buildRequest(url, getMethod(), useRequestCache)

is never executed?

wseemann committed:
Merge branch 'master' into okhttp-reuse-finetune

wseemann commented:
@pupitetris I didn't receive a response to my outstanding questions so I built off the concepts you were trying to introduce in this MR and integrated the changes myself [here](https://github.com/wseemann/Roku-ECP-Wrapper-Kotlin/commit/bd8041fbef6ea68412f0827772d05ca6476124a3). Thank you for your contribution!
